Improve efficiency and formatting of ITM output

kjbracey opened the pull request:
### Description

SerialWireOutput was outputting 1 character per 32-bit write to the ITM stimulus port. This is inefficient, and causes processing problems with some viewers due to them receiving 3 NUL bytes between each desired character.

Rework to allow us to be more efficient, and eliminate those NUL bytes:

* Retain existing mbed_itm_send() and clarify it's a single 32-bit write.
* Add new mbed_itm_send_block() that is appropriate for sending character data, and modify SerialWireOutput to use it.
* Move "wait for FIFO ready" check to before the write, rather than after.

One minor correction - FIFOREADY is a single bit of the register read. Don't interpret reserved bits.

Fixes #7373 

### Pull request type

    [ ] Fix
    [X] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

kjbracey committed:
Improve efficiency and formatting of ITM output

SerialWireOutput was outputting 1 character per 32-bit write to the
ITM stimulus port. This is inefficient, and causes processing problems
with some viewers due to them receiving 3 NUL bytes between each
desired character.

Rework to allow us to be more efficient, and eliminate those NUL bytes:

* Retain existing mbed_itm_send() and clarify it's a single 32-bit write.
* Add new mbed_itm_send_block() that is appropriate for sending
  character data, and modify SerialWireOutput to use it.
* Move "wait for FIFO ready" check to before the write, rather than
  after.

One minor correction - FIFOREADY is a single bit of the register read.
Don't interpret reserved bits.

kjbracey commented:
@RobMeades - something for you to test. I don't have any ITM boards handy.

kjbracey commented:
Revised a couple of times - now compiles, at least.

RobMeades commented:
Sweet.  Performing `printf("Hello world %d.\n", x);` in a loop now produces the following output in JLink SWO Viewer *and* in GDB's SWO console:

```
Hello world 0.
Hello world 1.
Hello world 2.
Hello world 3.
Hello world 4.
```

kjbracey commented:
I'd be interested to see if you could actually measure a performance boost. Or is the port so fast it doesn't really matter?

It may be that it's not really worth the extra code size of doing the 32-bit writes for the block operation.

kjbracey commented:
Fixed doxygen - no functional change.

kjbracey commented:
On efficiency, it's possible the C library doesn't end up doing anything more than single-character `write()` calls anyway. I believe IAR's C library can't be persuaded to do otherwise, and the other libraries may default to non-buffered and need to be explicitly switched to line- or fully-buffered.

You could perhaps test theoretical throughtput more directly with some manual `write(STDOUT_FILENO)`.

RobMeades commented:
For pure `printf()` output it appears to work quite fast enough: it didn't really work at all in any of the clients available to me before so I can't easily see what the performance improvement might be.  If I sit in a tight loop `printf()`ing 10 characters at a time I notice no loss in GDB and lines go past way faster than I can read; at a _very_ rough estimate 150 kbytes/s.

marcuschangarm reviewed (dismissed):
Looks great!

cmonr reviewed (dismissed):
Excited for the changes!

cmonr commented:
/morph build

cmonr commented:
@ARMmbed/mbed-os-maintainers Heads up that this PR adds a new function to HAL.
_However_, I've marked this as for a patch because 1) this is fixing behavior that should be working, and 2) the API addition does not regress user functionality.

kjbracey commented:
Sorry - need to correct this. The last commit ca45797 I added was wrong - we do still need the "enabled" checks.

Rereading documentation indicates that writing to the stimulus register is safe when disabled, but we will be getting "not ready" indications when we read, causing us to block indefinitely if disabled.

Removed that commit, leaving just the commit @RobMeades tested.

0xc0170 commented:
/morph build

cmonr commented:
Rebase is needed.

cmonr commented on hal/mbed_itm_api.c:
```diff
@@ -66,21 +70,65 @@ void mbed_itm_init(void)
+    /* Check if ITM and port is enabled */
+    if (((ITM->TCR & ITM_TCR_ITMENA_Msk) != 0UL) &&      /* ITM enabled */
+            ((ITM->TER & (1UL << port)) != 0UL)) {       /* ITM Port enabled */
+    {
```
Extra curly brace.

kjbracey commented:
/morph build
